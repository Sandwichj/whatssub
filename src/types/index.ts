import { DefaultTheme } from 'styled-components';

export interface User {
  displayName: string;
  age: number;
  job: string;
}

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
};

export interface ScreenProps {
  theme: DefaultTheme;
  changeTheme: () => void;
}
