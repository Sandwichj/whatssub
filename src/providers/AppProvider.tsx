import * as Font from 'expo-font';

import React, { useEffect, useReducer } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  useEffect(() => {
    Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpocaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpocaHanSans/SpoqaHanSans-Regular.ttf'),
    });
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
