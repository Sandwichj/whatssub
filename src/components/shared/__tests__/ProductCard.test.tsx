import 'react-native';
import * as React from 'react';
import ProductCard from '../ProductCard';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {
  render,
  act,
  RenderResult,
  fireEvent,
} from '@testing-library/react-native';
import { createTheme } from '../../../theme';
import { ThemeType } from '../../../types';
import { ThemeProvider } from 'styled-components/native';
import { AppProvider } from '../../../providers';

const component = (props?: any) => {
  return (
    <AppProvider>
      <ThemeProvider theme={createTheme(ThemeType.LIGHT)}>
        <ProductCard {...props}/>
      </ThemeProvider>
    </AppProvider>
  );
};

let rendered: renderer.ReactTestRendererJSON;
let testingLib: RenderResult;
// test for the container page in dom
describe('[ProductCard] ui rendering test', () => {
  it('should render outer component and snapshot matches', () => {
    act(() => {
      rendered = renderer.create(component()).toJSON();
    });
    expect(rendered).toMatchSnapshot();
  });

  it('should include button when "variant" is subscription', () => {
    const Component = component({ variant: 'subscription' });
    act(() => {
      rendered = renderer.create(Component).toJSON();
    });
    testingLib = render(Component); // Todo: add SwitchToggle testID
  });

  it('should include summary title text when product card is pressed', () => {
    const Component = component();
    testingLib = render(Component);
    act(() => {
      rendered = renderer.create(Component).toJSON();
      fireEvent.press(testingLib.getByTestId('productCard'));
    });
    testingLib.getByTestId('summaryTitleText');
  });
});

describe('[ProductCard] Interaction', () => {
  let testingLib: RenderResult;

  it('should simulate [notification Icon] click', () => {
    let cnt = 0;
    const Component = component({
      isNotificationEnable: true,
      onClickNotification: () => cnt++,
    });
    testingLib = render(Component);
    act(() => {
      fireEvent.press(testingLib.getByTestId('notiOffIcon'));
    });
    expect(cnt).toBe(1);
  });
});
