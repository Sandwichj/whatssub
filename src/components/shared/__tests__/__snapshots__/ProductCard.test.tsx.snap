// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProductCard] ui rendering test should render outer component and snapshot matches 1`] = `null`;
